compiler/noirc_frontend/src/parser/parser.rs

```diff
@@ -613,7 +613,22 @@ fn trait_implementation() -> impl NoirParser<TopLevelStatement> {
 }
 
 fn trait_implementation_body() -> impl NoirParser<Vec<TraitImplItem>> {
-    let function = function_definition(true).map(TraitImplItem::Function);
+    let function = function_definition(true)
+        .validate(|f, span, emit| {
+            if f.def().is_internal
+                || f.def().is_unconstrained
+                || f.def().is_open
+                || f.def.visibility != FunctionVisibility::Private
+            {
+                emit(ParserError::with_reason(ParserErrorReason::TraitImplFunctionModifiers, span));
+            }
+            f
+        })
+        .map(|mut f| {
+            // Trait impl functions are always public
+            f.def_mut().visibility = FunctionVisibility::Public;
+            TraitImplItem::Function(f)
+        });
```

Comment:
Might as well combine validate and map
```suggestion
        .validate(|mut f, span, emit| {
            if f.def().is_internal
                || f.def().is_unconstrained
                || f.def().is_open
                || f.def.visibility != FunctionVisibility::Private
            {
                emit(ParserError::with_reason(ParserErrorReason::TraitImplFunctionModifiers, span));
            }
            // Trait impl functions are always public
            f.def_mut().visibility = FunctionVisibility::Public;
            TraitImplItem::Function(f)
        });
```